File: frontend/src/utils/validators.ts
/**
 * File Validation Utilities for PDFMentor
 *
 * Provides client-side validation for PDF file uploads to ensure
 * security and proper file handling before sending to backend.
 */

// Maximum allowed file size: 1MB (in bytes)
const MAX_FILE_SIZE = 1 * 1024 * 1024; // 1MB

// Allowed MIME type for uploads
const ALLOWED_MIME_TYPE = 'application/pdf';

/**
 * Result of file validation
 */
interface ValidationResult {
  valid: boolean;      // Whether the file passed validation
  error?: string;      // Error message if validation failed
}

/**
 * Validates a PDF file before upload
 *
 * Checks:
 * - File type matches PDF MIME type
 * - File has .pdf extension
 * - File size is within limits
 * - File is not empty
 *
 * @param file - The File object to validate
 * @returns Validation result with success status and error message if any
 */
export const validatePDFFile = (file: File): ValidationResult => {
  // Check file type using MIME type
  if (file.type !== ALLOWED_MIME_TYPE) {
    return {
      valid: false,
      error: 'Only PDF files are allowed'
    };
  }

  // Check file extension as additional validation
  if (!file.name.toLowerCase().endsWith('.pdf')) {
    return {
      valid: false,
      error: 'File must have .pdf extension'
    };
  }

  // Check file size doesn't exceed maximum
  if (file.size > MAX_FILE_SIZE) {
    return {
      valid: false,
      error: `File size must be less than ${MAX_FILE_SIZE / 1024 / 1024}MB`
    };
  }

  // Check for empty file
  if (file.size === 0) {
    return {
      valid: false,
      error: 'File is empty'
    };
  }

  // All validations passed
  return { valid: true };
};

/**
 * Sanitizes a filename by removing potentially dangerous characters
 *
 * This is a basic sanitization - the backend will do more thorough
 * sanitization for security.
 *
 * @param filename - The original filename
 * @returns Sanitized filename safe for display
 */
export const sanitizeFilename = (filename: string): string => {
  // Remove path traversal attempts and special characters
  // Replace unsafe characters with underscores
  return filename.replace(/[^a-zA-Z0-9._-]/g, '_');
};

File: backend/app/utils/validators.py
"""
File validation utilities

Provides server-side validation for uploaded files to ensure security.
Validates file extensions, MIME types, and file sizes.
"""
import os
from werkzeug.utils import secure_filename
from flask import current_app

def allowed_file(filename: str) -> bool:
    """
    Check if file extension is allowed
    
    Args:
        filename: Name of the file to check
    
    Returns:
        True if file extension is in ALLOWED_EXTENSIONS, False otherwise
    """
    return '.' in filename and \
           filename.rsplit('.', 1)[1].lower() in current_app.config['ALLOWED_EXTENSIONS']

def validate_pdf_file(file) -> tuple[bool, str]:
    """
    Validate uploaded PDF file with multiple security checks

    Performs the following validations:
    1. File existence check
    2. Filename validation
    3. File extension check
    4. File size check
    5. PDF header validation (magic bytes)

    Args:
        file: FileStorage object from Flask request

    Returns:
        Tuple of (is_valid, error_message)
        - is_valid: True if file passes all validations
        - error_message: Description of error if validation fails
    """
    # Check if file exists
    if not file:
        return False, "No file provided"

    # Check filename
    if file.filename == '':
        return False, "No file selected"

    # Check file extension
    if not allowed_file(file.filename):
        return False, "Only PDF files are allowed"

    # Read first few bytes to check file size and magic number
    file.seek(0, os.SEEK_END)
    file_size = file.tell()
    file.seek(0)

    # Check file size
    if file_size == 0:
        return False, "File is empty"

    if file_size > current_app.config['MAX_FILE_SIZE']:
        max_mb = current_app.config['MAX_FILE_SIZE'] / (1024 * 1024)
        return False, f"File size exceeds {max_mb:.0f}MB limit"

    # Validate PDF header (magic bytes)
    # PDF files must start with %PDF-
    header = file.read(5)
    file.seek(0)

    if header != b'%PDF-':
        return False, "File is not a valid PDF"

    return True, ""

def sanitize_filename(filename: str) -> str:
    """
    Sanitize filename to prevent path traversal attacks
    
    Uses werkzeug's secure_filename which:
    - Removes path components (../, etc.)
    - Removes special characters
    - Converts to ASCII
    
    Args:
        filename: Original filename
    
    Returns:
        Sanitized filename safe for filesystem use
    """
    return secure_filename(filename)
